feat(extra-natives/five): DEFINE_DRY_VOLUME

HeadshotsOps opened the pull request:
This adds a new native that allows specifying a volume where water collision effects do not apply. This functionality is based on R*' way of disabling water collision under the casino.

This native allows volumes to be defined once and they will persist trough the whole game lifetime.

One question would be if it would be better to move the casino dry volume to this new collection of volumes too?

Disquse commented:
Hey, after looking a bit at your code it seems there's no way to remove those volumes? Maybe you can return array index as a handle so people could remove those volumes in scripts.

It seems there's no resetting code anywhere? So if player leave a server that used this native, he would also have those volumes on any other server that he join later, until he restart his game. You can use `OnShutdownSession` for example to clear volumes.

Also, `DEFINE_` sounds weird to be honest, there's no any single R* native with this prefix, would make more sense to use `CREATE_` or `ADD_` imo.

HeadshotsOps commented:
Hi, thanks for your review and useful feedback. 
I did indeed not think about resetting between sessions and I have also added a native to remove volumes now.
Also the array index is now returned when creating volumes.
The naming was changed to CREATE_ , it was indeed sounding a bit unusual.
I have implemented removals by marking the volumes as removed and setting their bounds to zero, this to keep previously defined volume handles valid instead of their index being shifted with a full removal from the vector. I'm not sure if this would be the most clean way.

austinsasko commented on ext/native-decls/RemoveDryVolume.md:
````diff
@@ -0,0 +1,15 @@
+```
+
+Removes a dry volume from the game session.
+See CREATE_DRY_VOLUME for more info
````
Should link back to the md file
```suggestion
See [CREATE_DRY_VOLUME.md](CREATE_DRY_VOLUME) for more info
```

blattersturm replied:
no, it shouldn't?!

blattersturm commented:
@Disquse @HeadshotsOps was there anything else blocking this?

Disquse commented on code/components/extra-natives-five/src/WaterExtraNatives.cpp:
```diff
@@ -0,0 +1,148 @@
+		int index = context.GetArgument<int>(0);
+		if (index > g_dryAreas.size() || index < 0)
+		{
+			trace("Attempting to delete an invalid dry volume handle (%d)", index);
```
Please add `\n`.

Disquse commented on code/components/extra-natives-five/src/WaterExtraNatives.cpp:
```diff
@@ -0,0 +1,148 @@
+	fx::ScriptEngine::RegisterNativeHandler("REMOVE_DRY_VOLUME", [=](fx::ScriptContext& context)
+	{
+		int index = context.GetArgument<int>(0);
+		if (index > g_dryAreas.size() || index < 0)
```
I believe there should be `index >= g_dryAreas.size()`.
You can remove dry volume indexed as `0` when you actually have no dry volumes created, this will cause client crash.
When you have at least 1 dry volume, if you call `REMOVE_DRY_VOLUME` for next index (i.e. for next volume, when it wasn't created yet) and then call `CREATE_DRY_VOLUME`, you will get crash.

Disquse commented on code/components/extra-natives-five/src/WaterExtraNatives.cpp:
```diff
@@ -0,0 +1,148 @@
+		context.GetArgument<float>(5),
+		});
+
+		std::vector<sDepthBound>::iterator v = std::find(g_dryAreas.begin(), g_dryAreas.end(), newVolume);
```
It seems AABB values needs to be limited, there's some weird side effects. If you're creating a dry volume with too high values (in my case I was testing -9999 to 9999), and your local ped is swimming, after creation this region some big part of the map would work like there's some water region (only physically). Wonder why "ped is swimming" part is very important for repro'ing this bug.
Here's repro video: https://gfycat.com/PlaintiveFaithfulAffenpinscher

HeadshotsOps replied:
It seems to be the case that creating a dry volume in an area where your ped is swimming, causes this. I cannot seem to figure out how to force the game to somehow reset the swimming state for now. However, dry volumes are likely created when the client joins since its a pretty static map feature. I don't see any usecase for removing water effects if a player is already in that pool anyways. Unless you think this is problematic enough to stop going into production?

Disquse reviewed (changes requested):
Found some issues, please take a look.

HeadshotsOps commented:
Thanks for the review, I will make the changes asap and look into that weird behavior

Disquse commented:
Also, please take care of possible memory leaks `new float[3]` needs to be destructed after using. Same as `newVolume` data, since you're storing a copy and not a reference in `g_dryAreas`.

HeadshotsOps commented:
Added @nikez ' edits #1612  and updated native decl. Anything else that needs to be changed?
